### PR TITLE
v6.0.0 - Tunneling/Transport/DataChannels Rewrite

### LINKED ISSUE #547: Tunneling Rewrite
...

### LINKED ISSUE #548: Datachannels Rewrite (Sunsetting WS)

### LINKED ISSUE #566: Memory leak data channels server _allowed
### Package Versions

Replace the values below using the output from `npm list storj`. Use ```npm list -g storj``` if installed globally.

```
Storj Share 3.0.1
Core Library 5.0.1
Protocol 0.10.0
```

Replace the values below using the output from `node --version`.

```
none
```

### Expected Behavior

Please describe the program's expected behavior. Include an example of your
usage code in the back ticks below if applicable.

I started my GUI with 40 GB used space and created a memory snapshot. At 80 GB and 100% used space I created a second memory snapshot and compared both snapshots. My node is not sending OFFER anymore because there is no free space available. If I still have something in my memory that has to do with OFFER it must be a memory leak.

### Actual Behavior

Please describe the program's actual behavior. Please include any stack traces
or log output in the back ticks below.

(https://github.com/Storj/core/blob/fc7877eced9022df403b94727bac2fe46456c4d6/lib/data-channels/server.js#L84)
_allowed contains 1681 objects and I have 1822 authorizing data channel on my logfile. (I didn't safe the logfile at the moment I created the memory snapshot)

Most of the objects are looking like this one.
```
_allowed: Object
0a0f5bcb273e34d144cd5a4385453d67f3ccd7b2: Object
client: null
contact: Object
address: "104.196.20.121"
hdIndex: 7729926
hdKey: "xpub6AHweYHAxk1EhJSBctQD1nLWPog6Sy2eTpKQLExR1hfzTyyZQWvU4EYNXv1NJN7GpLYXnDLt4PzN874g6zSjAQdFCHZN7U7nbYKYVDUzD42"
lastSeen: 1479912718070
nodeID: "c481bc0e556f1a89bf7cb620abc324d6fd384e29"
port: 8468
protocol: "0.10.0"
userAgent: "5.1.0"
expires: 1480102297398
hash: "c5b997eb528039a275f0acc5478e4f6504f8623d"
```

### Steps to Reproduce

Please include the steps the reproduce the issue, numbered below. Include as
much detail as possible.

1. Start GUI and start farming. Don't allocate all space. Farm only a few GB. (On windows it must be 32bit because 64bit memory snaphots can't be displayed with all details)
2. Wait a few minutes. The node needs time to send OFFER and open data channels. Better for the comparison later.
3. Create memory snapshot 1
4. Keep farming until the allocated space is full. Keep it running one additional hour to be sure GC had time to clean up everything
5. Create memory snapshot 2 and compere both snapshots.

### PR DESCRIPTION
**Pull request by bookchin**
* Closes #547 
* Closes #548 
* Closes #566 
* Refactors the transport adapter

> This PR completely eliminates **all** `DataChannel*` and `Tunnel*`, replacing them with simpler and more elegant solutions to remove technical debt and to yield a more maintainable code base.
> **It is also backwards incompatible.**

### CODE REVIEW

File: test/utils.unit.js [hunk 1]
```diff
@@ -383,4 +383,28 @@ describe('utils', function() {
 
   });
 
+  describe('#createShardDownloader', function() {
```
Comment (braydonf) [inline test/utils.unit.js]: We could use some more assertions that the url is formatted properly here.

File: test/network/transport.unit.js [hunk 2]
```diff
@@ -369,4 +377,233 @@ describe('Network/Transport', function() {
...
+        nodeID: utils.rmd160('nodeid')
+      }));
+      expect(receive.calledWithMatch(null)).to.equal(true);
+      done();
+    });
+
+    it('should receive the serialized message', function(done) {
+      var message = '{"id":"test","method":"PING","params":{}}';
+      var receive = function(data) {
+        expect(Buffer.isBuffer(data)).to.equal(true);
```
Comment (braydonf) [inline test/network/transport.unit.js]: We should also assert that data is the message as expected also here?

Comment (bookchin) [inline test/network/transport.unit.js]: doneski

**Comment by braydonf** [comment]
Finished code review, and looks good to me. 

Tested download/upload via https://github.com/bookchin/storj-integration/commit/970554527bea7b62a9b2babf9d3ec04b6edc1416

**Comment by bookchin** [comment]
Tunneling tested with hack.bookch.in:8080 environment

**Comment by braydonf** [comment]
Tested tunneling through hack.bookch.in

**Comment by bookchin** [comment]
:ship:

**Pull request merged**